### ISSUE TITLE
Pnumero dummy objective

### ISSUE THREAD

**Issue by eslickj** [Pyomo/pyomo#1217]
Pynumero adds a dummy objective to the model.  This could cause trouble.  In my case it just doesn't detect objectives that are not at the top level block, so I can end up with two objectives.  It could also cause trouble if you use pynumero on subblocks or want to use pynumero and add an objective later.

**Comment by santiagoropb** [comment]
If I recall correctly the dummy objective was necessary to avoid strange behavior with asl. The dummy objective was only added when no objective was present in the model. @carldlaird

**Comment by eslickj** [comment]
I guess the problem is that the way I use Pyomo sometimes I end up solving a problem, then adding an objective later.

**Comment by carldlaird** [comment]
PyNumero should probably match the Pyomo behavior. I will look at this with the current NLP API redesign going on.

**Comment by blnicho** [comment]
>  In my case it just doesn't detect objectives that are not at the top level block

@eslickj could you clarify this statement?

I agree with @carldlaird that we should match the behavior of Pyomo. We should be wary of modifying the model that the user hands to PyNumero unless we clone the model or "reset" the model to its original state by removing/deactivating the dummy objective function.

**Comment by eslickj** [comment]
@blnicho, I think with our IDAES structure we have a ConcreteModel on top then in that a Flowsheet.  As best I can remember at this point, when I add an objective to the Flowsheet, Pynumero thought I didn't have an objective and added one to the top level model.  Then, I had two, and everything broke down.

**Comment by blnicho** [comment]
Ok thanks. That looks like a bug with the existing implementation that is fixed in the new PyNumero PR that Carl opened by using `component_data_objects` instead of `component_map` to identify any existing Objective components

Dealing with your other use-case of solving a model with no objective function and then adding one later still needs to be addressed. I suspect the dummy objective function will still cause issues there.

**Comment by carldlaird** [comment]
I didn't know that Pyomo would work without an objective function - do we add a dummy objective somewhere in our interface?

**Comment by blnicho** [comment]
I suspect we do but I can't tell you where it happens in our code

**Comment by ghackebeil** [comment]
I’m not aware of any solver interfaces that add a dummy objective.

**Comment by carldlaird** [comment]
I have modified PyNumero to search for objectives differently (with component_data_objects and descend_into=True), so this should be fixed. Please see PR #1253. This should be  out of WIP once I fix the test failures.

**Comment by eslickj** [comment]
Can you just have the solver part add an objective if necessary and not have Pyomo do it at all?  That seems safest.  I've definitely run the PETSc solver with no objective, so a lot of stuff should work without one, although I guess it may reduce Pynumero's functionality. You should still be able to get the Jacobian at least.  

Maybe you could raise some exception for things that require an objective, but don't have one.  Then the user could add their own objective.  At least the user would have control, and it would be a bit safer.

**Comment by carldlaird** [comment]
Yes, we can (and should) do this. In fact, I put a comment in the code to this effect. However, this will require a change to the C-code side of PyNumero that I will not be able to do right away. At this point, I thought it better to ask the user to add a dummy objective rather than create one behind the scenes that they don't know about (or know how to deactivate later).
